### ISSUE TITLE
Programmatic way to update nix.conf

### ISSUE THREAD

**Issue by fzakaria** [NixOS/nix#3824]
**Is your feature request related to a problem? Please describe.**
I would love a concise way to specifically have people use my custom S3 binary cache by running a command to update their _nix.conf_ with a new **substituters** & **trusted-public-keys**

**Describe the solution you'd like**
```
nix option add-trusted-public-keys fmzakari-nixcache:BW+j5qekr4kkgv8SHRIdaQiDgVHQ5Z587kgjUrFdTec=
nix option add-substituters s3://fmzakari-nixcache
```

The concept behind this request is similar to Git's facility to even update it's config file.
it allows for guides or README to be followed with which to change Nix settings.

The example above has an _add_ & _remove_ flavor so that you can append & remove from the options that are lists.

**Describe alternatives you've considered**
I've considered taking the corresponding code from [cachix](https://github.com/cachix/cachix) and writing my own little utility.

**Comment by Ericson2314** [comment]
Most people use NixOS (or e.g. `nix-darwin`), where such a thing wouldn't work, and to be honest I think the philosophy carries over to where we don't like imperatively changing configurations.

I assume you have a project for which you've made a cachix cache? What you *really* want is a caching just for your project, and something like a project-only `.nix/conf` you can put right in the git repo.

The avoids the need for dubious changing of global settings.

**Comment by fzakaria** [comment]
You kind of hit the nail on the hammer i guess.
That definitely makes sense -- or if I can set the values in my `shell.nix`  so support for it in a per derivation setup.

**Comment by adrian-gierakowski** [comment]
@Ericson2314 could you please elaborate on how one would use a project-only `.nix/conf`? 
Particularly:
1. could I make various `nix` commands pick it up automatically?
2. would it work with [lorri](https://github.com/target/lorri)?

thanks!

**Comment by manveru** [comment]
I think you can already do that by combining `direnv` with setting `NIX_USER_CONF_FILES` to a path in your project.

**Comment by domenkozar** [comment]
Note that you first need #2149

**Comment by Ericson2314** [comment]
Yeah we'll need some way of having multiple config files. And then we'll need multiple trust maps with CA derivations to really get the most mileage out of it without security issues.

**Comment by zimbatm** [comment]
Wouldn't you want to override the substituters for a given project by default?

The settings that I usually want per project are:
* substituters: use the project's cache only
* trusted-public-keys: same
* sandbox: sometimes I want to use `relaxed` sandboxing to trade reproducibility for convenience.

**Comment by DrSensor** [comment]
related https://github.com/NixOS/nix/issues/2704

**Comment by burdiyan** [comment]
Oh, I just realized that nix-shell supports `--max-jobs` too, so one could pass it via direnv too, seems to be working. I set `use nix --max-jobs auto` in my `.envrc`.

**Comment by abhillman** [comment]
Something quite related here is bazel's _buildozer_ tool, which could form the basis of such a tool that @fzakaria mentions.

https://github.com/bazelbuild/buildtools/blob/master/buildozer/README.md#examples

`nix-editor`, which already exists, appears to implement some of this functionality. It would be great to have related functionality in `nix` itself. https://github.com/vlinkz/nix-editor

^ cc/ @vlinkz

I could also imagine a mode that shows a diff related to an intended change.

**Comment by abhillman** [comment]
One exceptionally common use-case that I would use such a tool for, for example, would be for example, trivially adding an entry to `environment.systemPackages = with pkgs; [` in my `configuration.nix` without having to open up vim, find the right line, etc.

**Comment by adrian-gierakowski** [comment]
> One exceptionally common use-case that I would use such a tool for, for example, would be for example, trivially adding an entry to `environment.systemPackages = with pkgs; [` in my `configuration.nix` without having to open up vim, find the right line, etc.

I believe this is out of scope as this issue is about editing [nix.conf](https://nixos.org/manual/nix/stable/command-ref/conf-file.html), not .nix files

**Comment by abhillman** [comment]
Ahh -- thanks for the clarification on this particular request. @adrian-gierakowski -- do you know of any related efforts or issues?

**Comment by adrian-gierakowski** [comment]
@abhillman https://github.com/vlinkz/nixos-conf-editor

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
